flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,238 @@
+import org.apache.iceberg.ReplacePartitions;
+import org.apache.iceberg.RowDelta;
+import org.apache.iceberg.SnapshotUpdate;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.WriteResult;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FilesCommitter implements Committer<FilesCommittable>, Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(FilesCommitter.class);
+  private static final String FLINK_JOB_ID = "flink.job-id";
+
+  // The max checkpoint id we've committed to iceberg table. As the flink's checkpoint is always increasing, so we could
+  // correctly commit all the data files whose checkpoint id is greater than the max committed one to iceberg table, for
+  // avoiding committing the same data files twice. This id will be attached to iceberg's meta when committing the
+  // iceberg transaction.
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient long maxCommittedCheckpointId;
+
+  private transient ExecutorService workerPool;
+
+  public FilesCommitter(TableLoader tableLoader, boolean replacePartitions, Map<String, String> snapshotProperties,
+                 int workerPoolSize) {
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+    this.maxCommittedCheckpointId = -1L;
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<FilesCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumRestored = 0;
+    int deleteFilesNumRestored = 0;
+
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> restored = Lists.newArrayList();
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+    long checkpointId = 0;
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+
+      if (request.getCommittable().checkpointId() > maxCommittedCheckpointId) {
+        if (request.getNumberOfRetries() > 0) {
```

Comment:
Why do we differentiate between the `restored` and the first time commits?

Reply:
In the restored case, this request may be from the pre task, it may be submitted successfully or failed. We use the flink job ID to obtain the checkpointid of the last successful commite and compare it with the checkpointid carried by the request to determine whether the data is successfully submitted.

Reply:
Are we sure that the `numberOfRetries` always reflect the state of the Iceberg table?
I think in both cases (retries or not) we should depend on the status of the Iceberg table, and only use the state of the Flink commits as a guide. If Flink says that something is committed, or the commit is failed, then we could be sure that the commit happened, but depending on missing information (no commit yet / no tries yet) could be erroneous.

Reply:
In the new version, I've given up using it only to reflect commit status. As mentioned above, I relied a little too much on Flink's fault-tolerance mechanism, and I went back to the latest way of using metadata information from iceberg table.